source/extensions/tracers/opentelemetry/tracer.cc

```diff
@@ -0,0 +1,144 @@
+                         const Upstream::HostDescriptionConstSharedPtr&) {
+  std::string trace_id_hex = absl::BytesToHexString(span_.trace_id());
+  std::string span_id_hex = absl::BytesToHexString(span_.span_id());
+  std::vector<uint8_t> trace_flags_vec{sampled()};
+  std::string trace_flags_hex = Hex::encode(trace_flags_vec);
+  std::string traceparent_header_value =
+      absl::StrCat(kDefaultVersion, "-", trace_id_hex, "-", span_id_hex, "-", trace_flags_hex);
+  // Set the traceparent in the trace_context.
+  trace_context.setByReferenceKey(kTraceParent, traceparent_header_value);
+}
+
+Tracer::Tracer(OpenTelemetryGrpcTraceExporterPtr exporter, Envoy::TimeSource& time_source,
+               Random::RandomGenerator& random, Runtime::Loader& runtime,
+               Event::Dispatcher& dispatcher, OpenTelemetryTracerStats tracing_stats)
+    : exporter_(std::move(exporter)), time_source_(time_source), random_(random), runtime_(runtime),
+      tracing_stats_(tracing_stats) {
+  flush_timer_ = dispatcher.createTimer([this]() -> void {
+    tracing_stats_.timer_flushed_.inc();
+    flushSpans();
+    enableTimer();
+  });
+  enableTimer();
+}
+
+void Tracer::enableTimer() {
+  const uint64_t flush_interval =
+      runtime_.snapshot().getInteger("tracing.opentelemetry.flush_interval_ms", 5000U);
+  flush_timer_->enableTimer(std::chrono::milliseconds(flush_interval));
+}
+
+void Tracer::flushSpans() {
+  ExportTraceServiceRequest request;
+  // A request consists of ResourceSpans.
+  ::opentelemetry::proto::trace::v1::ResourceSpans* resource_span = request.add_resource_spans();
+  ::opentelemetry::proto::trace::v1::InstrumentationLibrarySpans* instrumentation_library_span =
+      resource_span->add_instrumentation_library_spans();
+  for (const auto& pending_span : span_buffer_) {
+    (*instrumentation_library_span->add_spans()) = pending_span;
+  }
+  tracing_stats_.spans_sent_.add(span_buffer_.size());
+  if (!exporter_->log(request)) {
+    // TODO: should there be any sort of retry or reporting here?
+    ENVOY_LOG(trace, "Unsuccessful log request to OpenTelemetry trace collector.");
+  }
+  span_buffer_.clear();
+}
+
+void Tracer::sendSpan(::opentelemetry::proto::trace::v1::Span& span) {
+  span_buffer_.push_back(span);
+  const uint64_t min_flush_spans =
+      runtime_.snapshot().getInteger("tracing.opentelemetry.min_flush_spans", 5U);
+  if (span_buffer_.size() >= min_flush_spans) {
+    flushSpans();
+  }
+}
+
+Tracing::SpanPtr Tracer::startSpan(const Tracing::Config& config, const std::string& operation_name,
+                                   SystemTime start_time,
+                                   const Tracing::Decision tracing_decision) {
+  // Create an Tracers::OpenTelemetry::Span class that will contain the OTel span.
+  Span new_span = Span(config, operation_name, start_time, time_source_, *this);
+  new_span.setSampled(tracing_decision.traced);
+  uint64_t trace_id_high = random_.random();
```

Comment:
Is it possible to make it "less random"? i.e. 
hex timestamp (32b) + hex ms (16b) + hex instance's IP (32b) + some internal counter?